python/src/scipp/neutron/instrument_view.py

```diff
@@ -0,0 +1,464 @@
+
+# Other imports
+import numpy as np
+import ipywidgets as widgets
+import matplotlib.pyplot as plt
+from matplotlib import cm
+try:
+    import ipyvolume as ipv
+except ImportError:
+    ipv = None
+import IPython.display as disp
+from matplotlib.collections import PatchCollection
+from matplotlib.patches import Rectangle
+
+
+def instrument_view(scipp_obj=None, bins=None, masks=None, filename=None,
+                    figsize=None, aspect="equal", cmap=None, log=False,
+                    vmin=None, vmax=None, size=0.1, projection="3D",
+                    nan_color="#d3d3d3", continuous_update=True,
+                    dim=sc.Dim.Tof):
+    """
+    Plot a 2D or 3D view of the instrument.
+    A slider is also generated to navigate the time-of-flight dimension.
+
+    Example:
+
+    import scipp.neutron as sn
+    sample = sn.load(filename="PG3_4844_event.nxs")
+    sn.instrument_view(sample)
+    """
+
+    iv = InstrumentView(scipp_obj=scipp_obj, bins=bins, masks=masks,
+                        cmap=cmap, log=log, vmin=vmin, vmax=vmax,
+                        aspect=aspect, size=size, projection=projection,
+                        nan_color=nan_color, filename=filename,
+                        continuous_update=continuous_update, dim=dim)
+
+    return SciPlot(iv.members)
+
+
+class InstrumentView:
+
+    def __init__(self, scipp_obj=None, bins=None, masks=None, cmap=None,
+                 log=None, vmin=None, vmax=None, aspect=None, size=1,
+                 projection=None, nan_color=None, filename=None,
+                 continuous_update=None, dim=None):
+
+        self.fig2d = None
+        self.fig3d = None
+        self.scatter2d = None
+        self.scatter3d = None
+        self.outline = None
+        self.size = size
+        self.aspect = aspect
+        self.do_update = None
+        self.figurewidget = widgets.Output()
+        self.figure2d = False
+        self.figure3d = False
+        self.image = None
+        self.nan_color = nan_color
+        self.log = log
+        self.current_projection = None
+        self.dim = dim
+
+        self.data_arrays = {}
+        tp = type(scipp_obj)
+        if tp is sc.Dataset or tp is sc.DatasetProxy:
+            for key, var in sorted(scipp_obj):
+                if self.dim in var.dims:
+                    self.data_arrays[key] = var
+        elif tp is sc.DataArray or tp is sc.DataProxy:
+            self.data_arrays[scipp_obj.name] = scipp_obj
+        else:
+            raise RuntimeError("Unknown input type: {}. Allowed inputs "
+                               "are a Dataset or a DataArray (and their "
+                               "respective proxies).".format(tp))
+
+        # Get detector positions from either coordinates or labels
+        if sc.Dim.Position in scipp_obj.coords:
+            self.det_pos = np.array(scipp_obj.coords[sc.Dim.Position].values)
+        else:
+            self.det_pos = np.array(scipp_obj.labels["position"].values)
```

Comment:
You can use `sc.neutron.position(scipp_obj)`. I notice that the docs miss the link to this particular function.

Reply:
It is currently not possible to get `.values` from a `VariableConstProxy` which `sc.neutron.position()` is returning. This thus requires new non-const overloads for the `sc.neutron.position()` function, and I delay this to a later PR.